core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -181,8 +181,13 @@ abstract class AbstractFetcherThread(name: String,
       val fetchedEpochs = fetchEpochsFromLeader(epochRequests)
       //Ensure we hold a lock during truncation.
       inLock(partitionMapLock) {
-        //Check no leadership changes happened whilst we were unlocked, fetching epochs
-        val leaderEpochs = fetchedEpochs.filter { case (tp, _) => partitionStates.contains(tp) }
+        //Check no leadership and no leader epoch changes happened whilst we were unlocked, fetching epochs
+        val leaderEpochs = fetchedEpochs.filter { case (tp, _) =>
+          val curPartitionState = partitionStates.stateValue(tp)
+          val leaderEpochInRequest = epochRequests.get(tp).get.currentLeaderEpoch.get
```

Comment:
Perhaps no harm being a little more defensive here. At least perhaps we can ensure `tp` is contained in `epochRequests`?

Reply:
But we still throw an exception, right? I guess it's better to throw IllegalStateException with a descriptive message vs. NPE.

Reply:
Or we can log a warning and ignore that partition. Well, perhaps an exception with a nice message is preferable since this would suggest the remote broker gave us data for a partition that we didn't request.

Reply:
Yeah, generally Option.get should never be used in favour of a more descriptive error.

Reply:
Yeah, generally Option.get should never be used in favour of a more descriptive error.